File: web/src/models/portfolio.ts
import type { Instrument, Quote } from "./market";

export interface Holding {
  id: string;
  instrument: Instrument;
  quantity: number;
  avg_cost: number;
  market_value?: number | null;
  unrealized_pl?: number | null;
  unrealized_pl_pct?: number | null;
  weight?: number | null;
}

export interface PortfolioSummary {
  market_value: number;
  cost_basis: number;
  unrealized_pl: number;
  unrealized_pl_pct: number;
  diversification_score: number;
  sector_exposure: Record<string, number>;
}

export interface Portfolio {
  id: string;
  name: string;
  summary: PortfolioSummary;
  holdings: Holding[];
}

export interface WatchlistItem {
  instrument: Instrument;
  quote?: Quote | null;
  added_at?: string | null;
}

export interface Watchlist {
  id: string;
  name: string;
  items: WatchlistItem[];
}

File: backend/app/schemas/portfolio.py
from __future__ import annotations

import uuid

from pydantic import BaseModel, Field

from app.schemas.market import InstrumentOut


class HoldingOut(BaseModel):
    id: uuid.UUID
    instrument: InstrumentOut
    quantity: float
    avg_cost: float
    market_value: float | None = None
    unrealized_pl: float | None = None
    unrealized_pl_pct: float | None = None
    weight: float | None = None


class PortfolioSummary(BaseModel):
    market_value: float
    cost_basis: float
    unrealized_pl: float
    unrealized_pl_pct: float
    diversification_score: float
    sector_exposure: dict[str, float] = {}


class PortfolioOut(BaseModel):
    id: uuid.UUID
    name: str
    summary: PortfolioSummary
    holdings: list[HoldingOut]


class AddHoldingRequest(BaseModel):
    symbol: str
    quantity: float = Field(gt=0)
    avg_cost: float = Field(ge=0)


class UpdateHoldingRequest(BaseModel):
    quantity: float = Field(gt=0)
    avg_cost: float = Field(ge=0)
